servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/ByteToMessageDecoder.java

```diff
@@ -217,39 +195,30 @@ public void channelReadComplete(ChannelHandlerContext ctx) {
         ctx.fireChannelReadComplete();
     }
 
-    private void tryDiscardSomeReadBytes(ByteBufAllocator allocator) {
-        // Avoid using discardSomeReadBytes because this will modify the underlying storage of the ByteBuf.
-        // Modifying the underlying storage of the ByteBuf means that any slices or views on this data will become
-        // corrupted. This is particularly a problem when offloading user code to a non-EventLoop thread to avoid
-        // blocking the EventLoop. In this case the users code will be executed asynchronously, which maybe a slice
-        // of this cumulation, and if we use discardSomeReadBytes the user code will see corrupted data.
-        if (cumulation != null && cumulation.readerIndex() >= cumulation.capacity() >>> 1) {
-            cumulation = swapCumulation(cumulation, allocator);
-        }
-    }
-
     /**
-     * Swap the existing {@code cumulation} {@link ByteBuf} for a new {@link ByteBuf}. This method is called when a
-     * heuristic determines the amount of unused bytes is sufficiently high that a resize / defragmentation of the
-     * bytes from {@code cumulation} is beneficial.
+     * Swap the existing {@code cumulation} {@link ByteBuf} for a new {@link ByteBuf} and copy {@code in}. This method
+     * is called when a heuristic determines the amount of unused bytes is sufficiently high that a
+     * resize / defragmentation of the bytes from {@code cumulation} is beneficial.
      * <p>
      * {@link ByteBuf#discardReadBytes()} is generally avoided in this method because it changes the underlying data
      * structure. If others have slices of this {@link ByteBuf} their view on the data will become corrupted. This is
-     * commonly a problem when processing data asynchronously to avoid blocking the EventLoop thread.
-     *
+     * commonly a problem when processing data asynchronously to avoid blocking the {@link EventLoop} thread.
+     * @param alloc Used to allocate a new {@link ByteBuf} if necessary.
      * @param cumulation The {@link ByteBuf} that accumulates across socket read operations.
-     * @param allocator Used to allocate a new {@link ByteBuf} if necessary.
-     * @return the {@link ByteBuf} that is responsible for accumulated socket reads.
+     * @param in The bytes to copy.
+     * @return the result of the swap and copy operation.
      */
-    protected ByteBuf swapCumulation(ByteBuf cumulation, ByteBufAllocator allocator) {
+    protected ByteBuf swapAndCopyCumulation(final ByteBufAllocator alloc, final ByteBuf cumulation, final ByteBuf in) {
+        ByteBuf newCumulation = alloc.buffer(alloc.calculateNewCapacity(
+                cumulation.readableBytes() + in.readableBytes(), MAX_VALUE));
```

Comment:
Should we take care about `int` overflow here? Otherwise, users may get `IllegalArgumentException` from `calculateNewCapacity`

Reply:
if we overflow `int` then we wouldn't be able to accommodate all the data in a single buffer. We could attempt to fall back to a multi-buffer approach, or provide a pre-check for a more informative exception, but since this case is unlikely to occur I think the current "check in a single spot" approach is sufficient for now.